### PR TITLE
feat: adds inSpan and SpanCustomizer.

### PR DESCRIPTION
**Pull request by jcchavezs**
This PR finally tackles #115 and adds the SpanCustomizer interface, allowing to do further customizations in http tracing.

Most important code: https://github.com/openzipkin/zipkin-php/pull/128/files#diff-ced4309e6e2dfd1e487df57084f07566R233

Ping carefully @adriancole

### CODE REVIEW

File: src/Zipkin/Tracer.php [hunk 1]
```diff
@@ -220,6 +224,49 @@ public function nextSpan(?SamplingFlags $contextOrFlags = null): Span
         throw new RuntimeException('Context or flags for next span is invalid.');
     }
 
+    /**
+     * @param callable $fn
+     * @param array $args
+     * @param callable $before
+     * @param callable $after
+     */
+    public function inSpan($fn, array $args, ?callable $argsParser = null, ?callable $resultParser = null)
```
Comment (jcchavezs) [inline src/Zipkin/Tracer.php]: This is the important piece.

File: src/Zipkin/SpanCustomizer.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    public function tag(string $key, string $value): void;
+
+    /**
+     * Associates an event that explains latency with the current system time.
+     *
+     * @param string $value A short tag indicating the event, like "finagle.retry"
+     * @param int $timestamp
+     * @return void
+     * @see Annotations
+     */
```
Comment (adriancole) [inline src/Zipkin/SpanCustomizer.php]: explicit timestamp desired?

File: src/Zipkin/Tracer.php [hunk 3]
```diff
@@ -220,6 +225,81 @@ public function nextSpan(?SamplingFlags $contextOrFlags = null): Span
         throw new RuntimeException('Context or flags for next span is invalid.');
     }
 
+    /**
+     * @param callable $fn
```
Comment (adriancole) [inline src/Zipkin/Tracer.php]: is this like trace local? maybe add a code example

File: src/Zipkin/SpanCustomizer.php (hunk 2 above)
Comment (jcchavezs) [reply src/Zipkin/SpanCustomizer.php]: That is actually a very nice catch. It turns out only `RealSpan` was making this optional. I improved this typing and other typings in https://github.com/openzipkin/zipkin-php/pull/130/files#diff-107b1bf5b96774ca3d48e6e084485747R107 and https://github.com/openzipkin/zipkin-php/pull/130/files#diff-554d86e667418ae9f107329f98c58bc8R76

File: src/Zipkin/Tracer.php (hunk 3 above)
Comment (jcchavezs) [reply src/Zipkin/Tracer.php]: Sure thing.

Comment (jcchavezs) [reply src/Zipkin/Tracer.php]: What do you think about this? https://github.com/openzipkin/zipkin-php/pull/128/files#diff-ced4309e6e2dfd1e487df57084f07566R228-R265

File: src/Zipkin/Tracer.php [hunk 4]
```diff
@@ -220,6 +225,122 @@ public function nextSpan(?SamplingFlags $contextOrFlags = null): Span
...
+     * @param array $args
+     * @param string|null $name the name of the span
+     * @param callable|null $argsParser with signature function (SpanCustomizer $span, ?array $args = []): void
+     * @param callable|null $resultParser with signature
+     * function (SpanCustomizer $span, $output = null, ?Throwable $e): void
+     */
+    public function inSpan(
+        callable $fn,
+        array $args = [],
+        ?string $name = null,
```
Comment (jcchavezs) [inline src/Zipkin/Tracer.php]: One thing to answer here is whether we force the user to pass a span name or we infer it from the callable (like we do it now in https://github.com/openzipkin/zipkin-php/pull/128/files#diff-ced4309e6e2dfd1e487df57084f07566R320)

**Comment by jcchavezs** [comment]
Only thing missing here is to define the context extraction for event loop. I will probably hold that until I finish my work with the current instrumentation for `driftphp`.

Ping @mmoreram

**Comment by jcchavezs** [comment]
This is green: https://travis-ci.org/github/openzipkin/zipkin-php/builds/689046500

**Pull request merged**